src/plugin_output_processing/settings.py

```diff
@@ -44,18 +48,49 @@ def check_model(self):
         and can be called, it will be used as the default provider. If not, OpenAI will
         be used if the API key is set. Otherwise, the service will not start.
         """
-        for provider in [Ollama(), OpenAI()]:
-            if self.provider and provider.name != self.provider:
-                continue
-            if provider.available:
-                self.provider = provider.name
-                self.model = provider.get_model(self.model)
-                self.url = provider.url
-                logger.debug(
-                    f"Provider set to {self.provider} with model {self.model}."
+        if self.provider == OLLAMA_NAME or self.provider is None:
+            ollama = Ollama(self.model)
+            if ollama.available:
+                return self.set_provider(ollama)
+            if self.provider is not None:
+                logger.warning(
+                    f"Ollama provider not available, falling back to {OPENAI_NAME}."
+                )
+                self.provider = None
+
+        if self.provider == OPENAI_NAME or self.provider is None:
+            openai = OpenAI(self.model)
+            if openai.available:
+                return self.set_provider(openai)
+            # If OpenAI is set in the configuration but not available, ollama would not
+            # have been tried yet. So by recursing, we can try to use ollama as a fallback.
+            if self.provider is not None:
+                logger.warning(
+                    f"OpenAI provider not available, falling back to {OLLAMA_NAME}."
                 )
-                return self
+                self.provider = None
+                return self.check_model()
+
+        msg = "None of the providers are available."
+        logger.error(msg)
+        raise ProviderError(msg)
+
+    def set_provider(self, provider: Provider):
+        logger.info(f"Provider set to {provider.name} with model {provider.model}.")
+        self.provider = provider.name
+        self.model = provider.model
+        self.url = provider.url
+        return self
```

Comment:
```suggestion
```

With the while loop this piece of code is not repeated anymore so maybe we can remove this function.

Reply:
I have a different opinion on the `while` loop, so keeping it for now.